lib/Conversion/HWToBTOR2/HWToBTOR2.cpp

```diff
@@ -0,0 +1,817 @@
+  // btor2 does not have the concept of a wire. This means that wires in
+  // hw will simply create an alias for the operation that will point to
+  // the same LID as the original op.
+  // key: alias, value: original op
+  DenseMap<Operation *, Operation *> opAliasMap;
+  // Stores the LID of the associated input.
+  // This holds a similar function as the opLIDMap but keeps
+  // track of block argument index -> LID mappings
+  DenseMap<size_t, size_t> inputLIDs;
+  // Stores all of the register declaration ops.
+  // This allows for the emission of transition arcs for the regs
+  // to be deferred to the end of the pass.
+  // This is necessary, as we need to wait for the `next` operation to
+  // have been converted to btor2 before we can emit the transition.
+  SmallVector<Operation *> regOps;
+
+  // Set of often reused strings in btor2 emission (to avoid typos and enable
+  // auto-complete)
+  static constexpr StringLiteral sortStr = "sort";
+  static constexpr StringLiteral bitvecStr = "bitvec";
+  static constexpr StringLiteral inputStr = "input";
+  static constexpr StringLiteral resetStr = "reset";
+  static constexpr StringLiteral outputStr = "output";
+  static constexpr StringLiteral zeroStr = "zero";
+  static constexpr StringLiteral oneStr = "one";
+  static constexpr StringLiteral constStr = "const";
+  static constexpr StringLiteral constdStr = "constd";
+  static constexpr StringLiteral consthStr = "consth";
+  static constexpr StringLiteral sliceStr = "slice";
+  static constexpr StringLiteral uextStr = "uext";
+  static constexpr StringLiteral addStr = "add";
+  static constexpr StringLiteral subStr = "sub";
+  static constexpr StringLiteral mulStr = "mul";
+  static constexpr StringLiteral andStr = "and";
+  static constexpr StringLiteral orStr = "or";
+  static constexpr StringLiteral xorStr = "xor";
+  static constexpr StringLiteral sllStr = "sll";
+  static constexpr StringLiteral srlStr = "srl"; // a.k.a. unsigned right shift
+  static constexpr StringLiteral sraStr = "sra"; // a.k.a. signed right shift
+  static constexpr StringLiteral sdivStr = "sdiv";
+  static constexpr StringLiteral udivStr = "udiv";
+  static constexpr StringLiteral smodStr = "smod";
+  static constexpr StringLiteral concatStr = "concat";
+  static constexpr StringLiteral notStr = "not";
+  static constexpr StringLiteral neqStr = "neq";
+  static constexpr StringLiteral hwNeqStr = "ne";
+  static constexpr StringLiteral iteStr = "ite";
+  static constexpr StringLiteral impliesStr = "implies"; // logical implication
+  static constexpr StringLiteral stateStr = "state";     // Register state
+  static constexpr StringLiteral nextStr = "next"; // Register state transition
+  static constexpr StringLiteral badStr = "bad";
+  static constexpr StringLiteral constraintStr = "constraint";
+  static constexpr StringLiteral wsStr = " ";  // WhiteSpace
+  static constexpr StringLiteral nlStr = "\n"; // NewLine
```

Comment:
Nit: I generally like having stringly things factored out if they are reused. It looks like the rest of your code is really nicely factored though and most of these constants only have a single use. It might be easier to just inline those string constants where they are used. Same for whitespace and newline.

Reply:
I did it this way because I have a tendency for typos and having constants like this avoided that. Now that the code is complete, I can inline them

Reply:
Makes sense :slightly_smiling_face: